File: pset6/vigenere/vigenere.py
import sys
from cs50 import get_string

# Define a main function to allow returns


def main():
    # Return 1 if incorrect number of arguments provided
    if len(sys.argv) != 2:
        print("Please provide one command line argument only")
        sys.exit(1)

    # Return 1 if the key is not letters ony
    key = sys.argv[1]
    if not key.isalpha():
        print("Please provide only alphabetical characters")
        sys.exit(1)

    # Convert key to list of shifts
    key = key.lower()
    keylist = []
    for i in key:
        keylist.append(ord(i) - 97)

    # Get string of plain text from the user
    plaintext = get_string("plaintext: ")
    ciphertext = ""
    shiftpos = 0
    shiftlen = len(keylist)

    # For each i in plaintext:
    for i in plaintext:

        # If a letter:
        if i.isalpha():
            # Calculate the shift
            shift = keylist[shiftpos]
            # Check if shift needs to loop back
            if ord(i.lower()) + shift > 122:
                ciphertext += chr(ord(i) + shift - 26)
            else:
                ciphertext += chr(ord(i) + shift)
            # Increment shift position
            if shiftpos + 1 == shiftlen:
                shiftpos = 0
            else:
                shiftpos += 1
        # If not a letter append the symbol
        else:
            ciphertext += i

    # Print the ciphertext
    print(f"ciphertext: {ciphertext}")


if __name__ == "__main__":
    main()

File: pset2/vigenere/vigenere.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>

int main(int argc, string argv[])
{
    //Return 1 if argc != 2, initialise k if 2
    if (argc != 2)
    {
        printf("Please provide 1 input only\n");
        return 1;
    }

    string key = argv[1];

    //Test the key for appropriateness (only letters)
    for (int i = 0, n = strlen(key); i < n; i++)
    {
        if (isalpha(key[i]) == 0)
        {
            printf("Please provide only alphabetical characters only\n");
            return 1;
        }
    }

    //Request plaintext from user an initialise c, p & k strings
    string p = get_string("plaintext: ");
    string c = p;
    int ki = 0;

    int j = 0;

    //For each element in p, conduct the conversion once with an if statement
    for (int i = 0, n = strlen(p); i < n; i++)
    {
        //Calculate ki from key j position and add 1 to j if a letter is being evaluated
        if (isalpha(p[i]))
        {
            if (islower(key[j]))
            {
                ki = key[j] - 97;
            }
            else
            {
                ki = key[j] - 65;
            }

            if (j < strlen(key) - 1)
            {
                j++;
            }
            else
            {
                j = 0;
            }
        }

        //One way for lower
        if (islower(p[i]))
        {
            if (p[i] + ki > 'z')
            {
                c[i] = 'a' - 1 + (ki - ('z' - p[i]));
            }
            else
            {
                c[i] = p[i] + ki;
            }
        }
        //One way for upper
        else if (isupper(p[i]))
        {
            if (p[i] + ki > 'Z')
            {
                c[i] = 'A' - 1 + (ki - ('Z' - p[i]));
            }
            else
            {
                c[i] = p[i] + ki;
            }
        }

    }
    printf("ciphertext: %s\n", c);
}
